More warhead polish

reaperrr opened the pull request:
Not dependent on, but related to #18054 and kind of a dependency for the follow-up (if only to reduce its diff).

reaperrr committed:
DamageWarhead polish

Use BitSet.IsEmpty.

reaperrr committed:
Use args in FireClusterWarhead methods

Instead of directly passing damage modifiers.

pchote commented:
@penev92 was recently complaining that `TargetDamageWarhead` appeared to be redundant vs `SpreadDamageWarhead` with `Falloff: 100, 100`. This might be a good time to merge the two?

pchote commented on OpenRA.Mods.Common/Warheads/FireClusterWarhead.cs:
```diff
@@ -52,23 +52,22 @@ public override void DoImpact(Target target, WarheadArgs args)
 		}
 
-		void FireProjectileAtCell(Map map, Actor firedBy, Target target, CPos targetCell, IEnumerable<int> damageModifiers)
+		void FireProjectileAtCell(Map map, Actor firedBy, Target target, CPos targetCell, WarheadArgs whArgs)
```
Style nit: this is part of the public API, so IMO it would be cleaner to call this one `args` then use `projectileArgs` below.

penev92 commented:
Code looks good but I agree with both comments. If we're going to merge the two might as well do it now and rebase all the other PRs on top of this. Or remove the (seemingly redundant?) spread from `TargetDamageWarhead` so it isn't a copy of `SpreadDamageWarhead`.
(Asking for those to be merged is a huge scope creep for this PR so is just a something to think about.)

reaperrr commented:
> (Asking for those to be merged is a huge scope creep for this PR so is just a something to think about.)

That alone wouldn't be such a problem. The problem is that `HealthPercentageWarhead` is based on `TargetDamageWarhead`, and the former is the one that actually uses `Spread` in the default mods, so I'd still have to shift the TargetDamage code to HealthPercWH (or vice-versa + rename).
Which then begs the question, couldn't/shouldn't we merge that into `SpreadDamageWarhead` too, and just add a `DamageMode` enum with `Absolute` and `HealthPercentage` as options? That would add `Falloff` support to health percentage damage logic as well.

pchote commented:
One big advantage of keeping a separate warhead for each damage model is that it sets a convention that is compatible with downstream mods if they want to introduce their own types.

reaperrr commented:
OK, but I feel in that case it may be better to leave this PR as-is wrt `TargetDamageWarhead`. There might be downstream warheads inheriting it like `HealthPercentageWarhead` does. IIRC we only introduced `TargetDamageWarhead` since we wanted `DamageWarhead` to remain abstract, and moving down its spread logic to `HealthPercentage` would require downstream warheads that don't want to inherit `HealthPercentage` to duplicate the `TargetDamage.Spread` logic if they rely on it.

reaperrr commented:
Updated, fixed the API style nit and accounted for `HealthPercentageWarhead` in the update rule.

If we do decide to merge some warheads, I'd like to leave that for a follow-up.

penev92 commented:
My personal preference would be to keep them distinctly separate (as opposed to the current "separate but identical").
Also looks like you need to rebase?

pchote commented:
> If we do decide to merge some warheads, I'd like to leave that for a follow-up.

Sounds good to me. This sounds messier than I first assumed.

reaperrr committed:
D2k yaml comment removal

Naming the warhead for its purpose is cleaner
than comments, in my opinion.

reaperrr commented:
Updated, removed the `TargetDamageWH` stuff, I think that's best dealt with in one go in a dedicated PR, and since it's unrelated to my follow-ups it's not something that needs to happen asap, plus I don't want to delay these smaller polishes over it.

abcdefg30 reviewed (approved):
Ok.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/1af90d2c7e051443147611c6e898f15edcba5d41...eb7618beea2b8e5723075e16cad589a129d3d060)
